Automattic/woocommerce-payments
# Fix fatal error on checkout with older versions of WC Subscriptions

Fixes #2409.
Fixes #2385. 

#### Changes proposed in this Pull Request


# WC Subscriptions < 3.1 Compatibility

The error is described in much greater detail in #2409; but to sum it up, on WC Subscriptions v.3.0.12 the global function `wcs_get_order_type_cart_items`, which we use to check if a subscription renewal is currently present in a customer's cart, does not exist and causes a fatal error if called, exterminating the entire payments section of the checkout. This change uses the less destructive `wcs_cart_contains_renewal` instead, with a compatibility check before the function is called. 

Moreover I have included a refactor of the Subscriptions trait, breaking out some general functions from the original trait that can now be shared between payment gateways and payment methods. 





#### Testing instructions


1. Please pull these changes, as well as the latest version of the Payments Server, and enable UPE checkouts either through the WCPay Dev Tools or by adding the `_wcpay_feature_upe` feature flag to your `wp_options` table with a value of `1`. Additionally, please install and activate WC Subscriptions version >= 2.20 and <= 3.0.12. (I tested this with v. 3.0.11, FYI.)
2. Add an item from your store to cart and proceed to checkout.
3. Note how there are no errors on the checkout page.
4. Complete a checkout and confirm that nothing has exploded along the way.
5. Repeat the above with a subscription product, just to be safe.
6. Sigh with relief.


-------------------
